clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -109,78 +112,76 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                 // we have enough unallocated or pooled memory to immediately
                 // satisfy the request
                 freeUp(size);
+                ByteBuffer allocatedBuffer = allocateByteBuffer(size);
```

Comment:
Yeah, it's not pretty. However, reducing the concurrency of `BufferPool` in the common path is not desireable. We definitely need to handle OOMs correctly, but they are relatively rare and it's OK if that path is slower.